File: PE_3/PE_3.cpp
//
//  main.cpp
//  C_PE_3 (largest prime factor)
//
//  Created by Neil Marshall on 18/11/2015.
//  Copyright © 2015 Neil Marshall. All rights reserved.
//

#include <iostream>
#include <vector>

/**
    Returns a vector of prime factors for a given integer.
*/
std::vector<long> factorize(long n) {

    std::vector<long> factors;
    long i = 2;
    while (n > 1) {
        while (n % i == 0) {
            factors.push_back(i);
            n /= i;
        }
        ++i;
    }

    return factors;

}

int main() {

    const long n = 600851475143;  // note: factors are 71, 839, 1471 and 6857

    std::vector<long> factors = factorize(n);

    std::cout << "Number to be factorised - " << n << "\n";
    std::cout << "Prime factors:" << "\n";
    for (auto factor : factors) std::cout << factor << " ";
    std::cout << std::endl;

}

File: 003/PE_3.fs
(* function that returns prime factors of a given number *)
let rec factorise n =
    let rec check_factor n known_factors factor_to_check =
        if factor_to_check = n then
            known_factors @ [n]
        else
            if n % factor_to_check = 0UL then
                check_factor (n / factor_to_check) (known_factors @ [factor_to_check]) 2UL
            else
                check_factor n known_factors (factor_to_check + 1UL)
    check_factor n [] 2UL

printfn "Factors of %d: %A" 600851475143UL (factorise 600851475143UL)
